src/output.cpp

```diff
@@ -236,161 +236,256 @@ std::string Output::get_helper_error_msg(int func_id, int retcode) const
+          reinterpret_cast<const AsyncEvent::Strftime *>(value.data())
+              ->strftime_id,
+          reinterpret_cast<const AsyncEvent::Strftime *>(value.data())->nsecs);
+    }
+    case Type::mac_address: {
+      return bpftrace.resolve_mac_address(value.data());
+    }
+    case Type::cgroup_path: {
+      return bpftrace.resolve_cgroup_path(
+          reinterpret_cast<const AsyncEvent::CgroupPath *>(value.data())
+              ->cgroup_path_id,
+          reinterpret_cast<const AsyncEvent::CgroupPath *>(value.data())
+              ->cgroup_id);
+    }
+    case Type::strerror: {
+      return strerror(read_data<uint64_t>(value.data()));
+    }
+    case Type::pointer:
+    case Type::reference: {
+      std::ostringstream res;
+      res << "0x" << std::hex << read_data<uint64_t>(value.data());
+      return res.str();
+    }
+    case Type::none: {
+      return "";
+    }
+    case Type::voidtype:
+    case Type::hist:
+    case Type::lhist:
+    case Type::stack_mode:
+    case Type::stats:
+    case Type::timestamp_mode: {
+      LOG(BUG) << "Invalid value type: " << type;
     }
-    return std::to_string(read_data<uint64_t>(value.data()) / div);
-  } else if (type.IsProbeTy())
-    return bpftrace.resolve_probe(read_data<uint64_t>(value.data()));
-  else if (type.IsTimestampTy())
-    return bpftrace.resolve_timestamp(
-        reinterpret_cast<AsyncEvent::Strftime *>(value.data())->mode,
-        reinterpret_cast<AsyncEvent::Strftime *>(value.data())->strftime_id,
-        reinterpret_cast<AsyncEvent::Strftime *>(value.data())->nsecs);
-  else if (type.IsMacAddressTy())
-    return bpftrace.resolve_mac_address(value.data());
-  else if (type.IsCgroupPathTy())
-    return bpftrace.resolve_cgroup_path(
-        reinterpret_cast<AsyncEvent::CgroupPath *>(value.data())
-            ->cgroup_path_id,
-        reinterpret_cast<AsyncEvent::CgroupPath *>(value.data())->cgroup_id);
-  else if (type.IsStrerrorTy())
-    return strerror(read_data<uint64_t>(value.data()));
-  else if (type.IsPtrTy() || type.IsRefTy()) {
-    std::ostringstream res;
-    res << "0x" << std::hex << read_data<uint64_t>(value.data());
-    return res.str();
-  } else {
-    assert(type.IsNoneTy());
-    return "";
   }
+  return "";
+}
+
+std::string Output::map_key_str(BPFtrace &bpftrace,
+                                const SizedType &arg,
+                                const std::vector<uint8_t> &data) const
+{
+  std::ostringstream ptr;
+  switch (arg.GetTy()) {
+    case Type::integer:
+    case Type::kstack:
+    case Type::ustack:
+    case Type::timestamp:
+    case Type::ksym:
+    case Type::usym:
+    case Type::inet:
+    case Type::username:
+    case Type::probe:
+    case Type::string:
+    case Type::buffer:
+    case Type::pointer:
+    case Type::array:
+    case Type::mac_address:
+    case Type::record:
+      return value_to_str(bpftrace, arg, data, false, 1, true);
```

Comment:
Just checking: any reason for the last argument to always be true?

(if I'm understanding correctly, it doesn't make a difference what value it has)

Reply:
This is needed for json output mode e.g. for this program
`bpftrace -e 'BEGIN { @["hello"] = 1; }' -f json`

without setting this to true you get this:
`{"type": "map", "data": {"@": {"\"hello\"": 1}}}`

whereas you want this:
`{"type": "map", "data": {"@": {"hello": 1}}}`